p2p/profile_service.go

```diff
@@ -0,0 +1,218 @@
+	"github.com/qri-io/qri/repo/profile"
+)
+
+// ProfileProtocolID is the protocol id for the profile exchange service
+const ProfileProtocolID = "/qri/profile/0.1.0"
+
+// ProfileTimeout is the length of time we will wait for a response in a profile
+// exchange
+const ProfileTimeout = time.Minute * 2
+
+// ProfileService manages the profile exchange. This exchange should happen
+// whenever a node connects to a new peer
+type ProfileService struct {
+	Host    host.Host
+	Repo    repo.Repo
+	pub     event.Publisher
+	peersMu *sync.Mutex
+	peers   map[peer.ID]chan struct{}
+}
+
+// NewQriProfileService creates an profile exchange service and adds
+// the `ProfileHandler` to the host
+func NewQriProfileService(r repo.Repo, p event.Publisher) *ProfileService {
+	ps := &ProfileService{
+		// Host:    h,
+		Repo:    r,
+		pub:     p,
+		peersMu: &sync.Mutex{},
+		peers:   map[peer.ID]chan struct{}{},
+	}
+	// h.SetStreamHandler(ProfileProtocolID, ps.ProfileHandler)
+	return ps
+}
+
+// StartProfileService adds a host and a profile handler to the host
+// to the profile service
+func (ps *ProfileService) StartProfileService(h host.Host) {
+	ps.Host = h
+	h.SetStreamHandler(ProfileProtocolID, ps.ProfileHandler)
+}
+
+// ProfileHandler listens for profile requests
+// it sends it's node's profile on the given stream
+// whenever a request comes in
+func (ps *ProfileService) ProfileHandler(s network.Stream) {
+
+	var (
+		err error
+		pro *profile.Profile
+	)
+
+	ctx := context.Background()
+
+	p := s.Conn().RemotePeer()
+
+	defer func() {
+		helpers.FullClose(s)
+		if err != nil {
+			ps.pub.Publish(ctx, event.ETP2PProfileSendFailed, p)
+			return
+		}
+		ps.pub.Publish(ctx, event.ETP2PProfileSendSuccess, p)
+	}()
+
+	log.Debug("%s received a profile request from %s %s", ProfileProtocolID, p, s.Conn().RemoteMultiaddr())
+
+	if err := ps.pub.Publish(ctx, event.ETP2PProfileRequestReceived, p); err != nil {
+		log.Errorf("%s publishing event %s", ProfileProtocolID, err)
+	}
+
+	pro, err = ps.Repo.Profile()
+	if err != nil {
+		log.Errorf("%s error getting this node's profile: %s", ProfileProtocolID, err)
+		return
+	}
+
+	err = sendProfile(s, pro)
+	if err != nil {
+		log.Errorf("%s error sending profile to %s: %s", ProfileProtocolID, p, err)
+		return
+	}
+}
+
+// // ProfileResult is a result of a ping attempt, either an RTT or an error.
+// type ProfileResult struct {
+// 	Profile *profile.Profile
+// 	Error   error
+// 	Wait    chan struct{}
+// }
+
+// ProfileRequest requests a profile from a specific peer
+func (ps *ProfileService) ProfileRequest(ctx context.Context, p peer.ID) {
+	<-ps.ProfileWait(ctx, p)
+	return
+}
+
+// ProfileWait checks to see if a request to this peer is already in progress
+// if not, it sends a profile request to the peer
+// either way it returns a channel to wait on until the profile request has
+// been completed
+func (ps *ProfileService) ProfileWait(ctx context.Context, p peer.ID) <-chan struct{} {
```

Comment:
Does this need to be exported? Looks like  `ProfileService` users should be calling `ProfileRequest`